File: StriverA2Z/JavaScript&TypeScript/src/Sorting/mergeSort.ts
// Merge Sort Algorithm

// Time Complexity: O(n log n)
// Space Complexity: O(n)

function mergeSort(arr: number[], low: number, high: number) {
  if (low === high) return;
  let mid = Math.floor((low + high) / 2);
  mergeSort(arr, low, mid);
  mergeSort(arr, mid + 1, high);
  merge(arr, low, mid, high);
}

function merge(arr: number[], low: number, mid: number, high: number) {
  let temp = [];
  let left = low;
  let right = mid + 1;
  while (left <= mid && right <= high) {
    if (arr[left] <= arr[right]) {
      temp.push(arr[left]);
      left++;
    } else {
      temp.push(arr[right]);
      right++;
    }
  }
  while (left <= mid) {
    temp.push(arr[left]);
    left++;
  }
  while (right <= high) {
    temp.push(arr[right]);
    right++;
  }
  for (let i = low; i <= high; i++) {
    arr[i] = temp[i - low];
  }
}

let arr4 = [3, 1, 2, 4, 1, 5, 2, 6, 4];
console.log("Before Sorting: ", arr4);
mergeSort(arr4, 0, arr4.length - 1);
console.log("After Sorting: ", arr4);

File: StriverA2Z/Java/SortingAlgorithms/MergeSort.java
import java.util.ArrayList;

public class MergeSort {
    public static void main(String[] args) {
        int[] arr = {3, 1, 2, 4, 1, 5, 2, 6, 4};
        mergeSort(arr, 0, arr.length - 1);
        for (int x : arr) System.out.print(x + " ");
    }

    static void mergeSort(int[] arr, int low, int high) {
        if (low >= high) return;
        int mid = (low + high) / 2;
        mergeSort(arr, low, mid);
        mergeSort(arr, mid + 1, high);
        merge(arr, low, high, mid);
    }

    static void merge(int[] arr, int low, int high, int mid) {
        ArrayList<Integer> intList = new ArrayList<>();
        int left = low;
        int right = mid + 1;

        while (left <= mid && right <= high) {
            if (arr[left] <= arr[right]) {
                intList.add(arr[left]);
                left++;
            } else {
                intList.add(arr[right]);
                right++;
            }
        }
        while (left <= mid) {
            intList.add(arr[left]);
            left++;
        }
        while (right <= high) {
            intList.add(arr[right]);
            right++;
        }

        for (int i = low; i <= high; i++) {
            arr[i] = intList.get(i - low);
        }
    }
}
